File: tests/presentacion_browser.js
/* DOM real y código de producción; almacenamiento aislado por caso. */
(async () => {
    const igual = (a, b) => {
        if (JSON.stringify(a) !== JSON.stringify(b)) throw new Error(`${JSON.stringify(a)} != ${JSON.stringify(b)}`);
    };
    function seleccionar(w, selector, value) {
        const control = w.document.querySelector(selector);
        control.value = value;
        control.dispatchEvent(new w.Event("change", {bubbles:true}));
    }
    const decimal = w => seleccionar(w, "#numeric-mode", "decimal");
    const texto = (w, selector) => w.document.querySelector(selector).textContent;
    const casos = [
        ["Exacto inicial; precisión oculta y pasos cerrados", {}, w => {
            igual(texto(w, "#solucion"), "x1 = 1/3");
            igual(w.document.querySelector("#numeric-mode").value, "exacto");
            igual(w.document.querySelector("#numeric-precision").hidden, true);
            igual(w.document.querySelector("#pasos").open, false);
        }],
        ["Decimal coherente en soluciones, pasos, matriz y etiquetas", {}, w => {
            decimal(w);
            igual(texto(w, "#solucion"), "x1 ≈ 0.3333");
            igual(texto(w, "#mitad"), "x2 = 0.5");
            igual(texto(w, "#entero"), "x3 = 4");
            igual(texto(w, "#pasos code"), "F1 ≈ (0.3333)F1");
            igual([...w.document.querySelectorAll("td")].map(n => n.textContent), ["0.3333", "-0.3333"]);
            igual(w.document.querySelector("table").getAttribute("aria-label"), "Matriz con factor 0.3333");
            igual(w.document.querySelector("#entrada").value, "1/3");
        }],
        ["Las cuatro precisiones, sin peticiones ni apertura de pasos", {}, w => {
            decimal(w);
            for (const p of [2,4,6,8]) {
                seleccionar(w, "#numeric-precision", String(p));
                igual(texto(w, "#solucion"), "x1 ≈ 0." + "3".repeat(p));
                igual(texto(w, "[data-numeric-notice]").includes(`${p} decimales`), true);
            }
            igual(w.requests, 0);
            igual(w.document.querySelector("#pasos").open, false);
        }],
        ["Vuelta al exacto y limpieza de aviso", {}, w => {
            decimal(w); seleccionar(w,"#numeric-mode","exacto");
            igual(texto(w,"#solucion"),"x1 = 1/3");
            igual(texto(w,"[data-numeric-notice]"),"");
            igual(w.document.querySelector("table").getAttribute("aria-label"), "Matriz con factor 1/3");
        }],
        ["Preferencia decimal y precisión recuperadas", {values:{"pygebra-formato-numerico":"decimal","pygebra-precision-decimal":"6"}}, w => {
            igual(texto(w,"#solucion"),"x1 ≈ 0.333333");
            seleccionar(w,"#numeric-precision","8");
            igual(w.localStorage.getItem("pygebra-precision-decimal"),"8");
            igual(w.localStorage.getItem("pygebra-formato-numerico"),"decimal");
        }],
        ["Preferencias inválidas vuelven a los predeterminados", {values:{"pygebra-formato-numerico":"otro","pygebra-precision-decimal":"99"}}, w => {
            igual(w.document.querySelector("#numeric-mode").value,"exacto");
            igual(w.document.querySelector("#numeric-precision").value,"4");
        }],
        ["Sin localStorage: exacto inicial y cambios funcionales", {blocked:true}, w => {
            igual(texto(w,"#solucion"),"x1 = 1/3"); decimal(w);
            igual(texto(w,"#solucion"),"x1 ≈ 0.3333");
            w.document.querySelector("#theme-toggle").click();
            igual(w.document.documentElement.classList.contains("js"),true);
        }],
        ["Tema histórico migrado", {values:{"algebra-lineal-tema":"dark"}}, w => {
            igual(w.document.documentElement.dataset.theme,"dark");
            igual(w.localStorage.getItem("pygebra-tema"),"dark");
        }],
        ["Tema PyGebra tiene prioridad sobre el histórico", {values:{"pygebra-tema":"light","algebra-lineal-tema":"dark"}}, w => {
            igual(w.document.documentElement.dataset.theme,"light");
            w.document.querySelector("#theme-toggle").click();
            igual(w.localStorage.getItem("pygebra-tema"),"dark");
        }],
        ["Sin JavaScript: valores exactos y details nativos", {nojs:true}, w => {
            igual(texto(w,"#solucion"),"x1 = 1/3");
            igual(w.document.querySelector("[data-numeric-controls]").hidden,true);
            w.document.querySelector("summary").click();
            igual(w.document.querySelector("#pasos").open,true);
        }],
    ];
    let passed=0;
    for (const [name, seed, test] of casos) {
        window.seed=seed;
        const frame=document.createElement("iframe");
        if(seed.nojs) frame.setAttribute("sandbox","allow-same-origin");
        frame.src="/fixture";
        const loaded=new Promise(resolve => frame.onload=resolve);
        document.body.append(frame);
        await loaded;
        const item=document.createElement("li");
        try { await test(frame.contentWindow); passed++; item.textContent=`PASS · ${name}`; }
        catch(error) { item.textContent=`FAIL · ${name}: ${error.message}`; }
        document.querySelector("#resultados").append(item);
        frame.remove();
    }
    document.querySelector("#total").textContent=`${passed}/${casos.length} PASS`;
})();

File: tests/presentacion_browser.py
"""Regresiones en DOM real: uv run python -m tests.presentacion_browser.

Abrir http://127.0.0.1:8876/. Sin dependencias nuevas; sirve los componentes
y scripts de producción con valores exactos preparados por Django.
"""

import os
import re
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import urlsplit

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "frontend.web.algebra_web.settings")
import django
django.setup()

from django.template import Context, Template

ROOT = Path(__file__).resolve().parents[1]
APP = ROOT / "frontend/web/calculadora"


def fixture():
    contenido = Template('''{% load numeros %}
        <input id="entrada" value="1/3" aria-label="Entrada original">
        {% numeric_results %}<section id="resultado">
        {% include "calculadora/components/numeric_format.html" %}
        <code id="solucion">x1 = 1/3</code><code id="mitad">x2 = 1/2</code>
        <code id="entero">x3 = 4</code>
        <details id="pasos"><summary>Ver procedimiento</summary><code>F1 = (1/3)F1</code>
        <table aria-label="Matriz con factor 1/3"><tr><td>1/3</td><td>-1/3</td></tr></table></details>
        </section>{% endnumeric_results %}''').render(Context())
    base = (APP / "templates/calculadora/base.html").read_text(encoding="utf-8")
    tema_inicial = re.search(r"<script>(.*?)</script>", base, re.S).group(1)
    return f'''<!doctype html><html lang="es" data-theme="light"><meta charset="utf-8">
        <script>
        const seed = parent.seed || {{}};
        const memory = new Map(Object.entries(seed.values || {{}}));
        Object.defineProperty(window, "localStorage", {{value: {{
            getItem(key) {{ if(seed.blocked) throw new Error("sin almacenamiento"); return memory.get(key) ?? null; }},
            setItem(key, value) {{ if(seed.blocked) throw new Error("sin almacenamiento"); memory.set(key, value); }}
        }}}});
        window.requests = 0;
        window.fetch = () => {{ window.requests++; throw new Error("petición inesperada"); }};
        window.XMLHttpRequest = function() {{ window.requests++; throw new Error("petición inesperada"); }};
        {tema_inicial}
        </script><button id="theme-toggle"><span class="theme-toggle-text">Tema</span></button>
        {contenido}<script src="/numeros.js"></script><script src="/tema.js"></script></html>'''.encode()


class PaginaPruebas(BaseHTTPRequestHandler):
    def do_GET(self):
        recursos = {
            "/": ("text/html", b'<!doctype html><html lang="es"><meta charset="utf-8"><title>PyGebra: regresiones de presentacion</title><h1>Presentacion numerica y tema</h1><ol id="resultados"></ol><p id="total">Ejecutando...</p><script src="/pruebas.js"></script></html>'),
            "/fixture": ("text/html", fixture()),
            "/numeros.js": ("text/javascript", (APP / "static/calculadora/numeros.js").read_bytes()),
            "/tema.js": ("text/javascript", (APP / "static/calculadora/tema.js").read_bytes()),
            "/pruebas.js": ("text/javascript", (ROOT / "tests/presentacion_browser.js").read_bytes()),
        }
        recurso = recursos.get(urlsplit(self.path).path)
        if recurso is None:
            self.send_error(404)
            return
        tipo, contenido = recurso
        self.send_response(200)
        self.send_header("Content-Type", tipo + "; charset=utf-8")
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(contenido)


if __name__ == "__main__":
    print("Pruebas DOM: http://127.0.0.1:8876/", flush=True)
    ThreadingHTTPServer(("127.0.0.1", 8876), PaginaPruebas).serve_forever()
